sw/host/opentitanlib/src/bootstrap/legacy.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use mundane::hash::{Digest, Hasher, Sha256};
+use std::time::Duration;
+use zerocopy::AsBytes;
+
+use crate::bootstrap::{BootstrapOptions, UpdateProtocol};
+use crate::io::spi::{Target, Transfer};
+
+#[derive(AsBytes, Debug, Default)]
+#[repr(C)]
+struct FrameHeader {
+    hash: [u8; Frame::HASH_LEN],
+    frame_num: u32,
+    flash_offset: u32,
+}
+
+#[derive(AsBytes, Debug)]
+#[repr(C)]
+struct Frame {
+    header: FrameHeader,
+    data: [u8; Frame::DATA_LEN],
+}
+
+impl Default for Frame {
+    fn default() -> Self {
+        Frame {
+            header: Default::default(),
+            data: [0xff; Frame::DATA_LEN],
+        }
+    }
+}
+
+impl Frame {
+    const EOF: u32 = 0x8000_0000;
+    const DATA_LEN: usize = 2048 - std::mem::size_of::<FrameHeader>();
+    const HASH_LEN: usize = 32;
+
+    /// Computes the hash in the header.
+    fn header_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let frame = self.as_bytes();
+        let sha = Sha256::hash(&frame[Frame::HASH_LEN..]);
+        sha.bytes()
+    }
+
+    /// Computes the hash over the entire frame.
+    fn frame_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let sha = Sha256::hash(self.as_bytes());
+        let mut digest = sha.bytes();
+        // Touch up zeroes into ones.
+        for i in 0..digest.len() {
+            if digest[i] == 0 {
+                digest[i] = 1;
+            }
+        }
+        digest
+    }
+
+    /// Creates a sequence of frames based on a `payload` binary.
+    fn from_payload(payload: &[u8]) -> Vec<Frame> {
+        let mut frames = Vec::new();
+
+        let mut max_addr = (payload.len() + 3) / 4;
+        while max_addr > 0 && payload[(max_addr - 1) * 4..max_addr * 4] == [0xff, 0xff, 0xff, 0xff] {
+            max_addr -= 1;
+        }
+
+        let mut frameno = 0;
+        let mut addr = 0;
+        let mut last_frame_end_addr = 0;
+        while addr < max_addr {
+            // Check if a range of all 0xff bytes can be skipped.
+            let mut nonempty = addr;
+            while nonempty < max_addr {
+                if payload[nonempty * 4..(nonempty + 1) * 4] != [0xff, 0xff, 0xff, 0xff] {
+                    break;
+                }
+                nonempty += 1;
+            }
+            if nonempty >= max_addr {
+                frameno |= Self::EOF;
+            } else {
+                let skip_addr = nonempty & !(512 - 1);
+                if skip_addr > addr {
+                    if last_frame_end_addr == 0 {
+                        addr = skip_addr;
+                    } else if last_frame_end_addr & 31 != 0 {
+                        addr = skip_addr;
+                    }
+                }
```

Comment:
This logic seems concerned with where to resume flashing after skipping a range of 0xFFs.

I don't quite understand what is happening with the `last_frame_end_addr & 31` case.  I recall there is an internal 128-byte buffer in H1D's bootstrap implementation, but I don't recall why you would or would not want to skip ahead based on how many bytes you sent in the last frame.

Can you add some comments to clarify the conditions here?

Reply:
I have copied comments from the spiflash.cc logic that this is a transcription of.